src/backend/distributed/master/worker_node_manager.c

```diff
@@ -400,6 +402,60 @@ NodeIsPrimaryWorker(WorkerNode *node)
 }
 
 
+/*
+ * CanUseCoordinatorLocalTablesWithReferenceTables returns true if we
+ * are allowed to use coordinator local tables with reference tables
+ * for joining or defining foreign keys between them.
+ */
+bool
+CanUseCoordinatorLocalTablesWithReferenceTables()
```

Comment:
> 2) When I do the ALTER TABLE ... ADD CONSTRAINT in a transaction block, I get an error that doesn't make sense in this context:

Finally catched the root cause of this problem. The problem is when we try to add a foreign key constraint between a `ref_table` and a `local_table`, postgres first tries to determine if there any of the existing rows violates the foreign key to be defined. To perform this check, it tries to join those two tables. In that moment, `CitusExecutorRun` takes its turn and we fall into that error.

I will check for transaction blocks in `PlanAlterTableStmt` to prevent this error

Reply:
> 3. ALTER TABLE table_name ADD CONSTRAINT ... locks table_name in ShareRowExclusiveLock. See https://www.postgresql.org/docs/current/explicit-locking.html for what this means. …

To do this, I guess I should call `UpdateRefTableIfFKeyBetweenLocalAndReferenceTables` at the very beginning of the `PlanAlterTableStmt` as we acquire the lock on relation in 	`Oid leftRelationId = AlterTableLookupRelation(alterTableStatement, lockmode);` statement.

This function does also check if we have required permissions on the target relation